hibernate-core/src/test/java/org/hibernate/test/collection/map/EmbeddableIndexTest.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.test.collection.map;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import javax.persistence.Embeddable;
+import javax.persistence.EmbeddedId;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.ManyToOne;
+import javax.persistence.MapKey;
+import javax.persistence.OneToMany;
+
+import org.hibernate.testing.junit4.BaseCoreFunctionalTestCase;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @author Andrea Boriero
+ */
+public class EmbeddableIndexTest extends BaseCoreFunctionalTestCase {
+
+	@Override
+	protected Class<?>[] getAnnotatedClasses() {
+		return new Class[] { TheOne.class, TheMany.class };
+	}
+
+	@Before
+	public void setUp() {
+		inTransaction(
+				session -> {
+					TheOne one = new TheOne( "1" );
+					session.save( one );
+
+					TheMapKey theMapKey = new TheMapKey( one );
+					TheMany theMany = new TheMany( theMapKey );
+					session.save( theMany );
+
+					Map<TheMapKey, TheMany> map = new HashMap<>();
+					map.put( theMapKey, theMany );
+					one.setTheManys( map );
+				}
+		);
+	}
+
+	@Test
+	public void testIt() {
+		inSession(
+				session -> {
+					TheOne one = session.get( TheOne.class, "1" );
+					TheMapKey theMapKey = one.getTheManys().keySet().iterator().next();
+					assertThat( theMapKey, is( notNullValue() ) );
+					assertThat( theMapKey.getTheOne(), is( one ) );
```

Comment:
It might be more accurate to use `sameInstance()` instead of `is()` for the following reasons:
* `is()` is based on `equalsTo()` but it seems there is no `equalsTo()` and `hashcode()` overriding in the class;
* should we allow `theMapKey.getTheOne()` returns another instance other than `one`? Seems not, then `sameInstance()` might be more accurate to express your intention.

Reply:
Thanks @NathanQingyangXu fixed it